internal/connection-manager/manager.go

```diff
@@ -0,0 +1,306 @@
+package connectionmanager
+
+import (
+	"fmt"
+	"log/slog"
+	"strings"
+	"sync"
+	"time"
+
+	mgmtv1alpha1 "github.com/nucleuscloud/neosync/backend/gen/go/protos/mgmt/v1alpha1"
+)
+
+type ConnectionProvider[T any] interface {
+	GetConnectionClient(connectionConfig *mgmtv1alpha1.ConnectionConfig) (T, error)
+	CloseClientConnection(client T) error
+}
+
+type ConnectionManager[T any] struct {
+	connectionProvider ConnectionProvider[T]
+	config             *managerConfig
+
+	mu *sync.Mutex
+
+	// Map of groupId -> sessionId -> connectionIds
+	groupSessionMap map[string]map[string]map[string]struct{}
+
+	// Map of groupId -> connectionId -> connection
+	groupConnMap map[string]map[string]T
+
+	shutdown chan any
+}
+
+type ConnectionInput interface {
+	GetId() string
+	GetConnectionConfig() *mgmtv1alpha1.ConnectionConfig
+}
+
+type Interface[T any] interface {
+	GetConnection(session SessionInterface, connection ConnectionInput, logger *slog.Logger) (T, error)
+	ReleaseSession(session SessionInterface) bool
+	ReleaseSessionGroup(grouper SessionGroupInterface) bool
+	Shutdown()
+	Reaper()
+}
+
+var _ Interface[any] = &ConnectionManager[any]{}
+
+type managerConfig struct {
+	closeOnRelease bool
+	reapDuration   time.Duration
+}
+
+// When a session is closed, if no other sessions within that group exist, the connection will be closed.
+// Otherwise, the connection will not be closed until Shutdown() or a Repear() cycle occurs.
+func WithCloseOnRelease() ManagerOption {
+	return func(mc *managerConfig) {
+		mc.closeOnRelease = true
+	}
+}
+
+func WithReaperPoll(duration time.Duration) ManagerOption {
+	return func(mc *managerConfig) {
+		mc.reapDuration = duration
+	}
+}
+
+type ManagerOption func(*managerConfig)
+
+func NewConnectionManager[T any](
+	connectionProvider ConnectionProvider[T],
+	opts ...ManagerOption,
+) *ConnectionManager[T] {
+	cfg := &managerConfig{
+		reapDuration: 1 * time.Minute,
+	}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	return &ConnectionManager[T]{
+		connectionProvider: connectionProvider,
+		groupSessionMap:    map[string]map[string]map[string]struct{}{},
+		groupConnMap:       map[string]map[string]T{},
+		config:             cfg,
+		shutdown:           make(chan any),
+		mu:                 &sync.Mutex{},
+	}
+}
+
+func (c *ConnectionManager[T]) GetConnection(
+	session SessionInterface,
+	connection ConnectionInput,
```

Comment:
might be helpful to have a few debug logs here about when it finds an existing vs when it creates a new one

Reply:
Good callout. I updated the interface to require the logger and filled it out throughout with more messages.
This forced me to pass it down further anyways and replaced a few instances where I was using the global slog.